src/fragment-generation-utils.js

```diff
@@ -157,6 +159,66 @@ export const generateFragment = (selection, startTime = Date.now()) => {
   return {status: GenerateFragmentStatus.AMBIGUOUS};
 };
 
+/**
+ * Checks whether fragment generation can be attempted for a given range. This
+ * checks a handful of simple conditions: the range must be nonempty, not inside
+ * an <input>, etc. A true return is not a guarantee that fragment generation
+ * will succeed; instead, this is a way to quickly rule out generation in cases
+ * where a failure is predictable.
+ * @param {Range} range
+ * @return {boolean} - true if fragment generation may proceed; false otherwise.
+ */
+export const isValidRangeForFragmentGeneration = (range) => {
```

Comment:
I wonder if we should also use this check to "fail-fast" when actually trying to generate a link.

Reply:
It's a tradeoff. The downside is doing duplicate calculations during the generation algorithm, which also needs to run quickly (though it's not *as* time-constrained as the new function). 

Certainly in the case of Chrome, we'll be gating the generation algorithm on a prior successful call to `isValidRangeForFragmentGeneration`, but we can't necessarily rely on other users doing the same. I don't have a great sense of which to prioritize.

This is where a tool like `DCHECK` would be nice to have, but maybe that's my C++ brain coming through.

Reply:
Ah right, yeah the call will be gated anyways. NVM!